Boot time speed up & metrics

jorgecuesta opened the pull request:
## High-Level Summary
This update includes changes that allow for faster boot time and better overall performance & maintainability.

* Mesh bootstrap time has been greatly reduced to almost 0, which improves overall performance and makes it easier to maintain your key files.
* Mesh now supports hot reloading of keys and chains efficiently by deduplicating full nodes, which allows for faster updates and better overall performance.
* Code has been split into smaller files to make it easier to navigate and modify, which helps with both development and maintenance.
* The Relay, Relay Time, and Error metrics now use labels for chain id and name to differentiate between different chains and to help with identifying and categorizing the various metrics.
* Misc bug fixes such as connection re-use and invalid sessions which should increase the performance of your mesh and full node.

### Breaking Changes
This changes require you update both: Pocket Node & Mesh Node

1. Added `/v1/private/mesh/check` endpoint
2. Moved `/v1/mesh/health` to `/v1/private/mesh/health`
3. Removed `/v1/private/mesh/servicer`
4. Removed Prometheus metrics using chain id as part of the metric name
5. Removed `authtoken` query param from private mesh endpoints in favor of `Authorization: <token>` header.
    *  **_NOTE: No changes need on your side unless you have some tooling built over `/v1/private/mesh/*` endpoints_**

**_NOTE:_** 

- You must add the new private endpoints (`check`) and use the same code/image on both sides or your mesh node will not start.
- You could remove the `/v1/health` endpoint from your proxy.


### Configuration Changes
1. Added `node_check_interval` to configure the rate at a full Node is health checked.
2. Added `log_chain_request` and `log_chain_response` to avoid unnecessary debug logs.
3. Added `chains_name_map` and `remote_chains_name_map` to enhance exposed metrics with chain name.
4. Added `metrics_moniker` to help to identify a mesh node instance on metrics queries.
7. Added `metrics_report_interval` to configure the rate at metrics about workers are reported.
8. Added `<servicer|chain>_rpc_max_idle_connections`
9. Added `<servicer|chain>_rpc_max_conns_per_host`
10. Added `<servicer|chain>_rpc_max_idle_conns_per_host`
11. Added `chain_drop_connections`
12. Added `client_rpc_read_timeout`
13. Added `client_rpc_read_header_timeout`
14. Added `client_rpc_write_timeout`
15. Added `chain_request_path_cleanup`
16. Removed `hot_reload_interval` in favor of `keys_hot_reload_interval` and `chains_hot_reload_interval` as separated values
17. Rename worker pool options to specify a different set for Servicer or Metrics
18. Remove `rpc_timeout` in favor of: `client_rpc_timeout` and `chain_rpc_timeout`

### Core Changes
#### New:
1. New keys structure
2. Deduplication of nodes (reduce bandwidth usage and faster boot time)
3. Allow control RPS of mesh to a node (using servicer_max_workers) due to how workers are used now
4. Chains check. Check against the Pocket node if the chains id on the chain file matches the one on Pocket Node.
6. Added a Worker for metrics
7. Chains name maps (local or remote) to enhance metrics
8. Added endpoint to allow mesh node check against servicer node (health, chains, addresses)
9. Added sanitization of the URL before call chain due to errors observed on chains like Avax/DFK sending `\t` characters on the path.

#### Fix:
1. Invalid sessions error
2. HTTP connection drops due to avoiding read response body (decreases CPU utilization and bandwidth)
3. Handle a few cases where after an error it keeps going without interrupting the flow, invalidating a relay/session.
4. Update chains using /v1/private/mesh/updatechains endpoint

#### Rework:
1. Relay, Relay Time, and Error metrics now use labels for chain id and name
2. Split Keys and Chains hot reload interval to allow them to work independently.
3. Moved the `authtoken` query parameter used to call private method to be passed using the HTTP header `Authorization` to avoid it being leaked on logs.
4. Split code in files to help development and readability of it.
5. Bootstrap time reduce to almost 0

#### Dependencies:
1. Added [xsync](github.com/puzpuzpuz/xsync) to have better performance that the native for concurrent access.
2. Added [gojsonschema](github.com/xeipuuv/gojsonschema) to allow validate keys and chains name map format.
3. Added [golang-set](github.com/deckarep/golang-set/v2) to manage set of values in arrays faster and async safe
4. Bump version github.com/alitto/pond from v1.8.1 to v1.8.3
5. Bump version github.com/prometheus/client_golang from v1.11.0 to v1.11.1

### Bump Version
New version: RC-0.3.0

### Docs
1. Update mesh.doc
2. Added missing links to dockerhub
3. Added Grafana dashboard to consume the new metrics.
4. Updated rpc-spec.yaml

jorgecuesta committed:
Initial rework to speed up bootstrap times of mesh client.
Added different key format.
Refactor connectivity checks.
Refactor node/servicer internal structure of mesh to reduce amount of worker/cron instances.
Refactor chains/keys reload.

nodiesBlade commented on app/cmd/rpc/mesh.go:
```diff
@@ -2420,11 +2807,22 @@ func GetServicerMeshRoutes() Routes {
 	return routes
 }
 
+func stopNode(node *fullNode) {
```
Use  receiver function

```go
func (n fullNode) stop() {
  //
}
```

jorgecuesta replied:
got u, let me update it.

nodiesBlade commented on app/cmd/rpc/mesh.go:
```diff
@@ -589,67 +677,122 @@ func reloadChains(chainsPath string) {
+	removedNodes.Each(func(s string) bool {
+		if v, ok := nodesMap.LoadAndDelete(s); ok {
+			n := v.(*fullNode)
+			for i := range n.Servicers {
```
```go
for _, servicer := range n.Servicers
```

jorgecuesta replied:
will they still be a reference? fullNode.Servicers is []*servicer but I do it on that way to ensure they not a new instance of the same object.

jorgecuesta replied:
resolved with the new xsync and rework after it

nodiesBlade commented on app/cmd/rpc/mesh.go:
```diff
@@ -589,67 +677,122 @@ func reloadChains(chainsPath string) {
+	removedServicers.Each(func(address string) bool {
+		if v, ok := servicerMap.LoadAndDelete(address); ok {
+			s := v.(*servicer)
+			for i := range s.Node.Servicers {
```
what does this do?

jorgecuesta replied:
remove servicers deleted (they were on the mesh before reloading the file and with the new file they are gone) from those nodes where they were because an option is that user remove a key from a fullNode but not the entire fullNode, or even move to a new fullNode so that it will remove from that fullNode to an "orphan" map so if there is another fullNode that is new, it will check first on orphan to grab the previous servicer instance, avoiding the need to reload the service app session cache again from remote fullNode on the first relay or throw errors on success relay notifications.

jorgecuesta replied:
resolved with the new xsync and rework after it

nodiesBlade reviewed (commented):
Ah. I don't have much insights to add. File loading/configuration can only be done so concisely. As long as it works!

jorgecuesta committed:
Fixed minor issues.
Added FullNode worker dynamic resize on servicers change.
Updated servicers reload to only run the modification on maps when there is something new/removed.

jorgecuesta commented:
@PoktBlade more changes on this, please check the latest commit

jorgecuesta committed:
Replace sync.Map with xsync.Map to use Generic in favor of performance and better readability of the code without so many casts.
Refactor fullNode.Servicer to be a map instead of a slice.
Enhance a bit more the logs and bootstrap time information.

jorgecuesta committed:
Added worker metrics.
Added metrics config support.
Refactor code to split in files.
Bump pond version to 1.8.3 (patch).
Clean up the code.

jorgecuesta committed:
Update docs.
Update config to handle rpc timeout for different things like chains, client and pocket node calls with a different value.

jorgecuesta committed:
Fixed metrics and cyclic dependencies.

jorgecuesta committed:
Fixed a bug that disallow the ability to reuse http 1.x connections.

jorgecuesta committed:
Added support to Chains & Servicer http client options to be configurable by config file.

jorgecuesta committed:
Updated rpc-spec.yaml.
Ensure that http response body is read even on errored request to reuse connections.

jorgecuesta committed:
Fixed issue with /v1/private/mesh/updatechains endpoint.
Enhanced chains reload logs.
Enhanced startup logs.

jorgecuesta committed:
Added specialized metrics for mesh instead of reuse the pocket client.

jorgecuesta committed:
Removed dynamic capacity on servicer worker due to the need of handle so many edge cases and possible infinite goroutine spams.
Added name property to nodes as optional key, if not set use the hostname of the node url.
Added minWorker, maxWorker, maxCapacity to prometheus metrics collectors.
Refactor minWorker, maxWorker and maxCapacity option in config.
Bump default to a more real world value.
Updated docs.

jorgecuesta committed:
Updated docs.

jorgecuesta committed:
Added option to allow debug logs omit request/response payloads.

jorgecuesta committed:
Added json schema support to additional properties. Those will not impact on mesh code, but help node runners to keep internal track.

jorgecuesta committed:
Rework metrics to handle less metrics with more labels. Enhance group/queries on prometheus.
Added chains name map so those metrics could contain the chain name you wish.
ChainsNameMap could work with a local file or remote endpoint (GET)

jorgecuesta committed:
Set chains_name_map default as empty so avoid to read it and write an error log. Does not affect the code but is unnecessary.

jorgecuesta committed:
Added metrics_moniker.
Added status_type and status_code labels to error metrics.
Added internal, notify and chain error metrics.

jorgecuesta committed:
Fixed issue with error metrics

jorgecuesta committed:
Rename status_label to status_code.
Added metrics docs and basic geo-mesh grafana dashboard.

nodiesBlade commented on app/cmd/rpc/mesh/metrics.go:
```diff
@@ -0,0 +1,438 @@
+
+func getErrorLabelSignature() []string {
+	if app.GlobalMeshConfig.MetricsAttachServicerLabel {
+		return []string{InstanceMoniker, ServicerLabel, ChainIDLabel, ChainNameLabel, NotifyLabel, StatusTypeLabel, StatusCodeLabel}
```
create a base array like

baseLabels := []string{InstanceMoniker, ServicerLabel, ChainIDLabel, ChainNameLabel, NotifyLabel, StatusTypeLabel, StatusCodeLabel}

if(servicerLabel)
baseLabels = append(...)

jorgecuesta replied:
done

nodiesBlade commented on app/cmd/rpc/mesh/metrics.go:
```diff
@@ -0,0 +1,438 @@
+}
+
+func getLabelSignature() []string {
+	if app.GlobalMeshConfig.MetricsAttachServicerLabel {
```
sames goes for above for base array

jorgecuesta replied:
done

nodiesBlade commented on doc/specs/mesh.md:
````diff
@@ -77,64 +84,75 @@ This guide assume you already have a Servicer properly setup and running. If not
 ```
-3. Create Servicer private key file with following format into the path you set on `config.json`:
+3. Create Servicer private key file with one of the  following formats into the path you set on `config.json`:
+Fallback Format:
````
Hide old format and only show as a spoiler / drop down..

users should use new format

jorgecuesta replied:
done

nodiesBlade commented on doc/specs/mesh.md:
````diff
@@ -146,8 +164,1521 @@ This guide assume you already have a Servicer properly setup and running. If not
+
+This dashboard allow you to monitor your mesh nodes across regions, so to properly work every metric need to contains an extra label called `region`
+You can add it on your scrape job with the following rule:
+```yml
````
add  dropdown / spoiler / new file for this

jorgecuesta replied:
solved as a new file

jorgecuesta committed:
Moved authtoken to Authorization header to avoid the security token could be leaked on logs.
Moved Grafana Dashboard to a file to easily compare/copy from raw github files.
Updated mesh.md

jorgecuesta committed:
Enhance code based on blade suggestions.
Enhance docs based on blade suggestions.

jorgecuesta committed:
Added golang-set dependency link to docs.

jorgecuesta committed:
Added missing chain request log.

jorgecuesta committed:
Added support to configure more http client that handle mesh request and the one dispatching to chains.

jorgecuesta committed:
Fixed issue with log chain requests.

jorgecuesta committed:
Allow to remove non-printable characters from chain path

nodiesBlade commented on app/mesh.go:
```diff
@@ -36,6 +36,7 @@ type MeshConfig struct {
 	ChainRPCMaxConnsPerHost     int    `json:"chain_rpc_max_conns_per_host"`
 	ChainRPCMaxIdleConnsPerHost int    `json:"chain_rpc_max_idle_conns_per_host"`
 	ChainDropConnections        bool   `json:"chain_drop_connections"`
+	ChainRequestPathCleanup     bool   `json:"chain_request_path_cleanup"`
```
should be default = true?

jorgecuesta replied:
done

nodiesBlade commented:
## High Level Summary
This update includes changes that allow for faster boot time and better overall performance & maintainability.

* Mesh bootstrap time has been greatly reduced to almost 0, which improves overall performance and makes it easier to maintain your key files.
* Mesh now supports hot reloading of keys and chains efficiently by duplicating full nodes, which allows for faster updates and better overall performance.
* Code has been split into smaller files to make it easier to navigate and modify, which helps with both development and maintenance.
* The Relay, Relay Time, and Error metrics now use labels for chain id and name to differentiate between different chains and to help with identifying and categorizing the various metrics.
* Misc bug fixes such as connection re-use and invalid sessions which should increase the performance of your mesh and full node.

### Breaking Changes
This changes require you update both: Pocket Node & Mesh Node

1. Added `/v1/private/mesh/check` endpoint
2. Moved `/v1/mesh/health` to `/v1/private/mesh/health`
3. Removed `/v1/private/mesh/servicer`
4. Removed Prometheus metrics using chain id as part of the metric name

**_NOTE: You must add the new private endpoints (`check` & `health`), or your mesh node will not start._**


### Configuration Changes
1. Added `node_check_interval` to configure the rate at a full Node is health checked.
2. Added `log_chain_request` and `log_chain_response` to avoid unnecessary debug logs.
3. Added `chains_name_map` and `remote_chains_name_map` to enhance exposed metrics with chain name.
4. Added `metrics_moniker` to help to identify a mesh node instance on metrics queries.
6. Added `metrics_report_interval` to configure the rate at metrics about workers are reported.
7. Added support for Servicer and Chain http.Transport options:
8. <servicer|chain>_rpc_max_idle_connections
9. <servicer|chain>_rpc_max_conns_per_host
10. <servicer|chain>_rpc_max_idle_conns_per_host
11. Removed `hot_reload_interval` in favor of `keys_hot_reload_interval` and `chains_hot_reload_interval` as separated values
12. Rename worker pool options to specify a different set for Servicer or Metrics
13. Remove `rpc_timeout` in favor of: `client_rpc_timeout` and `chain_rpc_timeout`

### Core Changes
#### New:
1. New keys structure
2. Deduplication of nodes (reduce bandwidth usage and faster boot time)
3. Allow control RPS of mesh to a node (using servicer_max_workers) due to how workers are used now
4. Chains check. Check against the Pocket node if the chains id on the chain file matches the one on Pocket Node.
5. Added a Worker for metrics
6. Chains name maps (local or remote) to enhance metrics
7. Added endpoint to allow mesh node check against servicer node (health, chains, addresses)

#### Fix:
1. Invalid sessions error
2. HTTP connection drops due to avoiding read response body (decreases CPU utilization and bandwidth)
3. Handle a few cases where after an error it keeps going without interrupting the flow, invalidating a relay/session.
4. Update chains using /v1/private/mesh/updatechains endpoint

#### Rework:
1. Relay, Relay Time, and Error metrics now use labels for chain id and name
2. Keys hot reload
3. Chains hot reload
4. Moved the `authtoken` query parameter used to call private method to be passed using the HTTP header `Authorization` to avoid it being leaked on logs.
5. Split code in files to help development and readability of it.
6. Bootstrap time reduce to almost 0

#### Dependencies:
1. Added [xsync](github.com/puzpuzpuz/xsync) to have better performance that the native for concurrent access.
2. Added [gojsonschema](github.com/xeipuuv/gojsonschema) to allow validate keys and chains name map format.
3. Added [golang-set](github.com/deckarep/golang-set/v2) to manage set of values in arrays faster and async safe
4. Bump version github.com/alitto/pond from v1.8.1 to v1.8.3
5. Bump version github.com/prometheus/client_golang from v1.11.0 to v1.11.1

### Bump Version
New version: RC-0.3.0

### Docs
1. Update mesh.doc
2. Added missing links to dockerhub
3. Added Grafana dashboard to consume the new metrics.
4. Updated rpc-spec.yaml


--

I made some changes to the release notes, could you elaborate more on the rework of:

```
2. Keys hot reload
3. Chains hot reload
```

jorgecuesta committed:
Updated default values.

jorgecuesta committed:
Updated mesh.md

jorgecuesta committed:
Updated mesh.md and default value of chain_request_path_cleanup

nodiesBlade reviewed (approved):
LGTM.

I think this is an iterative step architecturally and code-wise to improve the quality of Mesh. This is obviously not its final form and plenty of more code quality & architectural improvements down the road.

amazing work @jorgecuesta , I imagine a future where the entire network is leveraging mesh in some form or way
